src/stan/lang/grammars/semantic_actions_def.cpp

```diff
@@ -894,114 +961,155 @@ namespace stan {
+      int lhs_num_dims = ca.var_type_.dims_.size();
+      int lhs_num_idxs = ca.var_dims_.dims_.size();
 
-      if (lhs_type.is_ill_formed()) {
-        error_msgs << "too many indexes for variable "
+      if (inferred_lhs_type.is_ill_formed()) {
+        error_msgs << "Too many indexes for variable"
                    << "; variable name = " << name
-                   << "; num dimensions given = " << num_index_dims
-                   << "; variable array dimensions = " << lhs_var_num_dims
+                   << "; specified indexes = " << lhs_num_idxs
+                   << "; variable array dimensions = " << lhs_num_dims
                    << std::endl;
         pass = false;
         return;
       }
-
-      base_expr_type lhs_base_type = lhs_type.base_type_;
-      base_expr_type rhs_base_type = a.expr_.expression_type().base_type_;
-      // allow int -> double promotion
-      bool types_compatible
-        = lhs_base_type == rhs_base_type
-        || (lhs_base_type == DOUBLE_T && rhs_base_type == INT_T);
-      if (!types_compatible) {
-        error_msgs << "base type mismatch in assignment"
-                   << "; variable name = "
-                   << a.var_dims_.name_
-                   << ", type = ";
-        write_base_expr_type(error_msgs, lhs_base_type);
-        error_msgs << "; right-hand side type=";
-        write_base_expr_type(error_msgs, rhs_base_type);
+      // no compound assign for array types  (std::vector)
+      std::string op_equals = ca.op_;
+      ca.op_ = op_equals.substr(0, op_equals.size()-1);
+      if (inferred_lhs_type.num_dims() > 0) {
+        error_msgs << "Cannot apply operator '" << op_equals
+                   << "' to array variable; variable name = "
+                   << name
+                   << ".";
         error_msgs << std::endl;
         pass = false;
         return;
       }
 
-      if (lhs_type.num_dims_ != a.expr_.expression_type().num_dims_) {
-        error_msgs << "dimension mismatch in assignment"
-                   << "; variable name = "
-                   << a.var_dims_.name_
-                   << ", num dimensions given = "
-                   << lhs_type.num_dims_
-                   << "; right-hand side dimensions = "
-                   << a.expr_.expression_type().num_dims_
+      expr_type lhs_type = inferred_lhs_type.type();
+      expr_type rhs_type = ca.expr_.expression_type();
+      if (lhs_type.is_primitive()
+          && boost::algorithm::starts_with(ca.op_, ".")) {
+        error_msgs << "Cannot apply element-wise operation to scalar"
+                   << "; compound operator is: " << op_equals
                    << std::endl;
         pass = false;
         return;
       }
+      if (lhs_type.is_primitive()
+          && rhs_type.is_primitive()
+          && (lhs_type == DOUBLE_T || lhs_type == rhs_type)) {
+        // done checking <prim> <op>= <prim>
+        pass = true;
+        return;
+      }
 
+      bool types_compatible =
```

Comment:
operators like `=` should always start lines, not end them.  You dno't need any of those parens, either, because `&&` binds tighter than `||` and `==` binds tighter than both.